DSD-1609: MultiSelect v2

gonuguntla opened the pull request:
Fixes JIRA ticket [DSD-1609](https://jira.nypl.org/browse/DSD-1609)

## This PR does the following:

- Adds the `MultiSelect` v2 component.
- Adds the `useMultiSelect` hook.
- Updates the `Accordion` component to include the `ariaLabel` and `buttonInteractionRef` props.

## How has this been tested?
Locally, Storybook

<!--- Please describe in detail how you tested your changes. -->

## Accessibility concerns or updates

<!--- Describe any accessibility concerns or updates that were made that should be known. -->

-

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.

### Front End Review:

<!--- This step is done AFTER creating a PR. -->
<!--- Vercel creates a static Storybook preview URL once the PR is created. -->
<!--- That preview URL is added by Vercel as a comment. -->

- [ ] Review the Vercel preview deployment once it is ready.

Commit:
DSD-1609: Enabled multi-select and implemented v2 version

Commit:
DSD-1609: Enabled multi-select and implemented v2 version

gonuguntla committed:
Merge branch 'development' into DSD-1609/multiselect-v2

Commit:
DSD-1609: multiSelectDialog using Accordian

Commit:
DSD-1609: added accordian inline styles

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -77,29 +94,43 @@ export const MultiSelect = chakra(
+        showAll = false,
         type,
-        width = "default",
+        width = "fitContent",
```
The default should be "full" according to the spec doc.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -10,30 +10,47 @@ export interface MultiSelectItem {
   /** The action to perform for clear/reset button of MultiSelect. */
   onClear?: () => void;
+
+  showLabel: boolean;
+  /** Boolean value used to toggle the visibility of the Show All button
+   * The default value is false. */
+  showAll: boolean;
```
`showLabel` and `showAll` are not props in the spec doc.

EdwinGuzman replied:
Why did you have to add them?

gonuguntla replied:
In the user story first another document was there I followed that. Again new document updated. In the new document showLabel, Show All is not there.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -77,29 +94,43 @@ export const MultiSelect = chakra(
       } = props;
 
       const commonProps = {
+        helperText,
```
This is ok for now. Once you remove the `ListBox` variant, you shouldn't have a need for this object and can pass down the values directly to the `MultiSelectDialog` component.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -25,30 +24,73 @@ export const MultiSelectDialog = chakra(
         onMixedStateChange,
         selectedItems,
+        showLabel,
+        showAll,
```
You are not using `showAll` in this file.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -25,30 +24,73 @@ export const MultiSelectDialog = chakra(
         onClear,
         onMixedStateChange,
         selectedItems,
+        showLabel,
```
You are not using `showLabel` in this file. You can remove it.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+            placeholder="Search for options"
+            onChange={handleSearchKeyword}
+            showHelperInvalidText={true}
+            showLabel
```
This should be hard coded to `showLabel={false}`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+
+      const showSearchInputBox = () => {
+        if (isSearchable) {
+          return <TextInput
```
`labelText` is missing. According to the spec doc, it should be `Search ${buttonText}`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+            isClearable={true}
+            isClearableCallback={clearSearchKeyword}
+            isRequired
+            placeholder="Search for options"
```
The `placeholder` should be `Search ${buttonText}`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectMenuButton.tsx:
```diff
@@ -40,22 +31,14 @@ const grow = keyframes`
   HTMLButtonElement,
   MultiSelectMenuButtonProps
->((props, ref?) => {
+>((props) => {
```
You still need the ref declared even if you don't use it. The `?` makes it optional. When removed, it logs this warning:
![Screen Shot 2023-12-12 at 11 43 46 AM](https://github.com/NYPL/nypl-design-system/assets/1280564/92b3d085-db84-43f1-81a1-d8066ba52f0b)

EdwinGuzman commented on src/components/MultiSelect/MultiSelectMenuButton.tsx:
```diff
@@ -65,81 +48,38 @@ const MultiSelectMenuButton = forwardRef<
     isOpen,
-    hasSelectedItems: getSelectedItemsCount,
+    hasSelectedItems:
+      getSelectedItemsCount === undefined ? 0 : getSelectedItemsCount,
```
in `multiSelectMenuButton.ts`, the expected value for `hasSelectedItems` is a boolean. In this check, the value can be either the number `0` or a string.

You can leave this as it was before as `hasSelectedItems: getSelectedItemsCount` since it is undefined (false) by default and the string gives it a truthy value. You can also do
`hasSelectedItems: !!getSelectedItemsCount` to make sure it's a boolean value.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectMenuButton.tsx:
```diff
@@ -65,81 +48,38 @@ const MultiSelectMenuButton = forwardRef<
-          __css={styles.selectedItemsCountButton}
-        >
-          <Box as="span" verticalAlign="text-bottom">
+        <>
```
This is based on the previous implementation but it should get updated. It is currently
```
  return (
    <>
      {getSelectedItemsCount && (
        <>
          <Button
   ...
```

but it can be simplified to:

```
return getSelectedItemsCount ? (
    <Button
      ....
    >
      ...
    </Button>
  ) : null;
```

You don't need the extra react fragment wrappers since this component is returning only one parent element.

EdwinGuzman commented on src/theme/components/multiSelectMenuButton.ts:
```diff
@@ -6,6 +6,7 @@ const MultiSelectMenuButton = {
     "menuButton",
     "selectedItemsCountButton",
     "toggleIcon",
+    "menuClearButton",
```
This is not needed. I'm a bit confused why it was added but perhaps you wanted to declare styles for the clear button?

EdwinGuzman commented on src/theme/components/multiSelectDialog.ts:
```diff
@@ -0,0 +1,101 @@
+import { defaultElementSizes } from "./global";
+
+const MultiSelectDialog = {
```
None of these style are being used. If you want to use these styles, you have to configure it in the component you want to use them in.

I'm assuming this is for `MultiSelectDialog`. In the file, on line 88, this is set up as `const styles = useMultiStyleConfig("MultiSelect", {`.

This should use this object name: `const styles = useMultiStyleConfig("MultiSelectDialog", {`.

It's a bit off if you use `"MultiSelectDialog"` so you may need to fix this file if you want to use these styles.

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -128,6 +135,32 @@ const MultiSelect = {
+        md: "200px",
+      },
+    },
+    hideAccordian: {
```
You're not using this style so you can remove it.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -25,30 +24,73 @@ export const MultiSelectDialog = chakra(
+
+      React.useEffect(() => {
+        const calculateListHeight = () => {
+          if (listOverflow === "scroll") {
+            setViewAllLabel("View all");
+            setListHeight("260px");
+            setListItemsCount(defaultItemsVisible);
```
`listOverflow` does not get updated in this component based on state so you can set these as default when they are declared rather than here.
For example, for `setViewAllLabel("View all");`, this would be better on line 52 as
```
const [viewAllLabel, setViewAllLabel] = useState<string>(listOverflow === "scroll" ? "View all" : "");
```

The same is true for `setListHeight`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -25,30 +24,73 @@ export const MultiSelectDialog = chakra(
+          }
+        };
+
+        calculateListHeight();
```
Because you only call this function in this `useEffect`, you can remove it and call the code inside `calculateListHeight` directly. In other words, remove the function and this line but keep the code inside of it:
```
useEffect(() => {
          if (listOverflow === "scroll") {
            setViewAllLabel("View all");
            setListHeight("260px");
            setListItemsCount(defaultItemsVisible);
          } else {
            setViewAllLabel("View less");
            setListHeight("");
            setListItemsCount(items.length);
          }

        if (listItemsCount === defaultItemsVisible) {
          displayDefaultItems();
        } else {
          setItemsList(items);
        }
      }, []);
```

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -25,30 +24,73 @@ export const MultiSelectDialog = chakra(
+        } else {
+          setItemsList(items);
+        }
+      }, []);
```
When you use `useEffect` with an empty dependency array, then you are only calling it once when it renders. I'm not sure if this is what you want. If it is, then you probably don't need the `useEffect` and need to refactor your code. Otherwise, you need to add the variables that are used inside the `useEffect` into this array.

This is a screenshot of the linter giving a warning:
<img width="1125" alt="Screen Shot 2023-12-12 at 12 14 27 PM" src="https://github.com/NYPL/nypl-design-system/assets/1280564/c39ca5b4-d8e1-465a-8b31-494fbc0e2f37">

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -25,30 +24,73 @@ export const MultiSelectDialog = chakra(
+      }, []);
+
+      // Sets the selected items count on the menu button.
+      let getSelectedItemsCount;
```
`getSelectedItemsCount` is not used anywhere and only declared once. You can refactor this to:
```
const getSelectedItemsCount = !!(selectedItems[id]?.items.length || 0);
```

A few reasons why:
- the value is a string but is being used on line 92 to compare to undefined. The result is a boolean.
- the string value itself is not being used so you can convert it to a boolean using the double exclamation mark `!!`.
- If the `selectedItems` is undefined, then the default is `0` which will convert to `false`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+                  // Hide the remaining children
+                  setDefaultChildCheckboxes((prev) => {
+                    // Assuming 'prev' is an array, and we want to update it by adding 'childItem'
+                    return [...prev, { ...childItem, display: "none" }];
```
I don't understand what this is trying to do. When I log this, I get this which seems wrong:
![Screen Shot 2023-12-12 at 12 26 32 PM](https://github.com/NYPL/nypl-design-system/assets/1280564/e0760072-4f30-467e-9f0a-f360708856f9)

Notice that "none" is being add individually as letters in the array.

EdwinGuzman replied:
There's a bigger issue. `defaultChildCheckboxes` is currently defined as a string. You have it as:
```
const [defaultChildCheckboxes, setDefaultChildCheckboxes] = useState("none");
```

When you pass `defaultChildCheckboxes` to the `Checkbox` component, you're passing it as `display={defaultChildCheckboxes}`. This is fine.

The problem is that here you are updating `defaultChildCheckboxes` to an array instead of a string. So this needs to be reworked.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+        if (itemsList.length > 0) {
+          return (
+            <Text
+              onClick={() => viewAllItems()}
```
You should not call `onClick` functions on divs or p elements. The `Text` component renders a `p` element. Instead, this should be a button.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+      const itemsNotFound = () => {
+        if (itemsList.length === 0) {
+          return (
+            <Text id="text-btn" marginTop="xs" isItalic="true">
```
`isItalic` expects a boolean value, not a string.
<img width="934" alt="Screen Shot 2023-12-12 at 12 35 10 PM" src="https://github.com/NYPL/nypl-design-system/assets/1280564/3c45e926-ba73-4e94-ba78-9f527a62072a">

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+
+      const buttonTextLabel = () => {
+        return (
+          <Text __css={styles.accordianLabel} title={buttonText}>
```
Did you mean to use `title` as the prop?

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+          ref={mergedRefs}
+          __css={styles}
+          {...rest}
+          onKeyPress={onKeyPress}
```
There's a problem with this. Users expect to use the space bar to check/uncheck a checkbox. With this set here, you cannot do that. Instead, the small button with the number of items selected is triggered.

EdwinGuzman replied:
Consider moving this functionality inside of `MultiSelectMenuButton`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectDialog.tsx:
```diff
@@ -113,121 +156,299 @@ export const MultiSelectDialog = chakra(
+            accordionData={[
+              {
+                accordionType: "default",
+                label: buttonTextLabel(),
```
We expect `label` to be a string. So the `buttonTextLabel` function can return just a string.

EdwinGuzman reviewed (commented):
This is a first pass at reviewing the updated `MultiSelect` component. Let me know if you have any questions.

Commit:
DSD-1609: Addressed some code review comments

Commit:
DSD-1609: Fixed lint issues

Commit:
DSD-1609: Removed onclick on outside functionality for multi select

Commit:
DSD-1609: updated multi-select test cases

gonuguntla committed:
Merge branch 'development' into DSD-1609/multiselect-v2

Commit:
Fixed filterbar and multiselect group test cases

Commit:
commented temp remove code for filterbar and multi-select

Commit:
commented temp remove code for filterbar and multi-select

Commit:
Skipped failed tets cases for filterbar

Commit:
updated multi-select snapshorts

Commit:
added test cases for MUltiselect

gonuguntla committed:
Merge branch 'development' into DSD-1609/multiselect-v2

Commit:
updated snapshot of multi-select

Commit:
updated snapshot of multi-select

Commit:
Merge branch 'DSD-1609/multiselect-v2' of https://github.com/NYPL/nypl-design-system into DSD-1609/multiselect-v2

Commit:
Implemented multi-select checkboxes disabled functionality

Commit:
updated multi-select documentation

Commit:
updated multi-select documentation

Commit:
Update multi-select documentation

Commit:
updated test cases

Commit:
updated test cases

Commit:
updated test cases

Commit:
updated documentation

Commit:
updated documentation

Commit:
With search input field example documentation

Commit:
With search input field example documentation

Commit:
updated latest snapshot

Commit:
updated latest snapshot

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -27,86 +28,142 @@ import Link from "../Link/Link";
+  code={`
+<MultiSelect
+  id="subject"
+  labelText="Subject Specialities"
```
Is this code example mean to be what is rendered under "Component Props"? If so, then the title should be "MultiSelect" and you can add a "..." after the "Art History" object to show that you can have more objects in the array.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -27,86 +28,142 @@ import Link from "../Link/Link";
+  code={`
+<MultiSelect
+  id="subject"
+  labelText="Subject Specialities"
```
Again, this is fine and great to show as an example, but it should match what gets rendered and in the example above, there is no "Parent option". It is set as "Design" so update this code example.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -27,86 +28,142 @@ import Link from "../Link/Link";
+    },
+    {
+      id: "parent_option",
+      name: "Parent option",
```
Again, match the rendered example.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -27,86 +28,142 @@ import Link from "../Link/Link";
+/>
 
+## With Search Input Field Example
+When isSearchable is set to true, it enables the input field to search the items.
```
update to 
```
`isSearchable`
```

so the prop name displays better.

EdwinGuzman replied:
Same with "true".

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -27,86 +28,142 @@ import Link from "../Link/Link";
   id="subject"
   labelText="Subject Specialities"
-  type="listbox"
+  isSearchable="true"
```
This prop is a boolean so this should be `isSearchable={true}`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -117,24 +174,36 @@ The expected data structure is an array of objects, in the following format:
 />
 
-If you are using `type = 'dialog'`, you can also add a second level of child
+## Using the items prop
```
This section should be moved higher in the doc. Move it below the "Accessibility" section and before the "With Children Items Example".

Doing this, you explain the general way that the `items` prop can be used and then you go into the various properties within each object, such as `children` and `isDisabled`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -272,8 +342,10 @@ function MultiSelectControlledExample() {
       items={items}
-      type="listbox"
       selectedItems={selectedItems}
+      helperText="Multi Select Helper Text"
```
Remove "labelText" from this example.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -114,17 +188,16 @@ const meta: Meta<typeof MultiSelect> = {
+    items: items,
+    isSearchable: false,
+    defaultItemsVisible: 5,
+    isDefaultOpen: false,
```
List _all_ the props alphabetically. If they don't have a value, set it to `undefined`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -133,13 +206,17 @@ export const ListboxControls: Story = {
     id: "multiselect-dialog",
-    labelText: "MultiSelect Dialog",
-    type: "dialog",
+    buttonText: "MultiSelect",
```
We don't want this `args` object for any story except the "withControls" story. So you can remove this args object and declare these values in the `MultiSelectStory` component.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -149,43 +226,52 @@ export const DialogControls: Story = {
-    }
-  }, [actionName, selectedItems]);
+export const withDisabledItemsExample: Story = {
+  args: {
```
Same idea here, remove `args` and define the values in the component in the render property.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -149,43 +226,52 @@ export const DialogControls: Story = {
-    />
-  );
+export const withSearchInputFieldExample: Story = {
+  args: {
```
Same idea here, remove args and define the values in the component in the render property.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -149,43 +226,52 @@ export const DialogControls: Story = {
   const multiSelectId = args.id;
+  const isSearchable = args.isSearchable;
+  const isDefaultOpen = args.isDefaultOpen;
+  const itemsArray = args.items;
```
Nitpick: you can update this to:
```
const MultiSelectStory = ({ id, isSearchable, isDefaultOpen, items }) => {
```

That way you don't have to pull in args and define each property as `const itemsArray = args.items;`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -68,29 +57,115 @@ const items = [
-        "The action to perform for save/apply button of multiselect. <br /> `onApply: () => void;`",
+        "Value used to view the all items for the MultiSelect component",
+      control: "radio",
+      options: ["scroll", "expand"],
```
Pull in the `multiSelectListOverflowArray` array and use it here.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -101,11 +176,10 @@ const meta: Meta<typeof MultiSelect> = {
       description: "Value used to set the width for the MultiSelect component",
       control: "radio",
-      options: ["default", "fitContent", "full"],
+      options: ["fitContent", "full"],
```
Pull in the `multiSelectWidthsArray` array and use it here.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+  listOverflow?: multiSelectListOverflowTypes;
   /** The label text rendered within the MultiSelect. */
-  labelText: string;
+  buttonText: string;
```
If possible, order the props alphabetically.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
   id: string;
   /** Set the default open or closed state of the Multiselect. */
-  isDefaultOpen?: boolean;
+  isDefaultOpen: boolean;
```
This should be an optional prop.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+  isDefaultOpen: boolean;
+  /** Boolean value used to enable the component's search functionality
+   * The default value is false. */
+  isSearchable: boolean;
```
This should be an optional prop.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
   /** Boolean value used to control how the MultiSelect component will render within the page and interact with other DOM elements.
    * The default value is false. */
-  isBlockElement?: boolean;
+  isBlockElement: boolean;
```
This should be an optional prop.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+  buttonText: string;
+  /** Numeric value to set the maximum number of checkbox items
+   * The default value is 5. */
+  defaultItemsVisible: number;
```
This should be an optional prop.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
-  /** The type of MultiSelect that will be rendered. */
-  type: "listbox" | "dialog";
+  /** The action to perform on the checkbox's onChange function.  */
+  onChange?: MultiSelectOnChange;
```
It seems you're missing to implement `onChangeSearch` and `onClearSearch`. The spec doc says "MVP Req. NO" so I'm assuming it's okay to not implement this now, is that right @bigfishdesign13 ?

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
- * selections. Two variants of the MultiSelect component are offered, each with
- * slightly different functionality and requirements.  Because of these
- * differences, the two variants are broken out in separate stories below.
+ * selections. The MultiSelect allows for an optional set of child checkboxes to be passed,
+ * which makes the “parent” checkbox function as a check/uncheck all toggle.
+ * If all of the children checkboxes are checked, the parent isChecked prop will be true.
+ * If only some of the child checkboxes are checked, the parent checkbox will have a isIndeterminate state,
+ * that can be checked or unchecked.
```
This description comes up twice in the doc, once in the beginning and then again in the "With Children Items Example". I think you can remove this text here and add  a more general description of the component.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
-        selectedItems,
-        type,
+      // Control the open or closed state of the MultiSelect.
+      const [isOpen] = useState(isDefaultOpen);
```
`useState` is meant to be use for variables that get updated dynamically. If you don't need the `setIsOpen` function, then you don't need `useState` at all. Just using `isDefaultOpen` should work.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+      const [isOpen] = useState(isDefaultOpen);
+      const [itemsList, setItemsList] = useState(items);
+      const [viewAllLabel, setViewAllLabel] = useState("View all");
+      const [listHeight, setListHeight] = useState("215px");
```
Don't use magic numbers. So define variables for the default 0, "215px", and "275px" with meaningful name so we know _why_ those values are being applied.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+          }
+          setItemsList(items);
+        }
+      }, [listOverflow, items]);
```
Github and the linter tells you that you need additional dependencies in the array:
<img width="970" alt="Screen Shot 2024-01-18 at 6 32 49 PM" src="https://github.com/NYPL/nypl-design-system/assets/1280564/c668c976-ef80-474f-93b3-3aa5cb933749">

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+            setViewAllLabel("View less");
+          }
+        }
+      }, [listOverflow, listItemsCount, items, defaultItemsVisible]);
```
This is missing `displayDefaultitems` in the array.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+      }, [listOverflow, listItemsCount, items, defaultItemsVisible]);
+
+      // Sets the selected items count on the menu button.
+      let getSelectedItemsCount: any;
```
This should be `: boolean`.

EdwinGuzman replied:
This should now be `: number`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+
+      // Sets the selected items count on the menu button.
+      let getSelectedItemsCount: any;
+      if (selectedItems[id]?.items.length > 0) {
+        getSelectedItemsCount = !!(selectedItems[id]?.items.length || 0);
+      }
```
This seems unnecessary. If the length is 0, then it is false. Can you try `const getSelectedItemsCount = !!(selectedItems[id]?.items.length || 0);` to remove the "if" statement?

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+        item: MultiSelectItem
+      ): boolean => {
+        // Update isDisabled for the parent item
+        item.isDisabled = isAllChecked(multiSelectId, item);
```
What is this meant to do? `isAllChecked` is checking if the children are checked and not disabled. This value then gets overridden by the if/else statement below.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+        setItemsList(filteredItems);
+      };
+
+      const clearSearchKeyword = () => {
```
Can you add a description to this function? For example, when do you need to reset the list of items when the search is cleared?

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+        let count = 0;
+
+        for (let i = 0; i < items.length && count < defaultItemsVisible; i++) {
+          const currentItem = { ...items[i] };
```
I think you can do `const currentItem = items[i];` to keep it simpler.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+            const showIndeterminateChilds = isIndeterminate(id, currentItem);
+            count +=
+              !showChilds || !showIndeterminateChilds
+                ? currentItem.children?.length
+                : items[i].children.length;
```
This is confusing. Both `currentItem` and `items[i]` are the same so regardless of what is happening, `count` will always get the length count which will be the same number.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+              marginBottom="s"
+            />
+          );
+        }
```
Return `null` for the base case.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+              onChange={onChangeSearch}
+              showLabel={false}
+              showRequiredLabel={false}
+              step={1}
```
You don't need `step` for a text input.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+              labelText={`Search ${buttonText}`}
+              isClearable={true}
+              isClearableCallback={clearSearchKeyword}
+              isRequired
```
You don't need this and it's not in the spec doc.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+              showLabel={false}
+              showRequiredLabel={false}
+              step={1}
+              textInputType="default"
```
Remove this.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+                layout="column"
+                isFullWidth
+                isRequired={false}
+                labelText=""
```
This is a required prop. If you don't want to display it, set `showLabel` to false.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+                name="multi-select-checkbox-group"
+              >
+                {itemsList.map((item: MultiSelectItem) => (
+                  <React.Fragment key={item.id}>
```
This is why you are getting the warning message: "NYPL Reservoir CheckboxGroup: Only `Checkbox` components are allowed as children."

The `React.Fragment` is a component when you add the key, so CheckboxGroup thinks you're passing something that is not a `Checkbox`.

What you can do is pass an array as a child to render to any React component, such as `[<Checkbox key={1} />, <Checkbox key={2} />, <Checkbox key={3} />, ...etc]` and this should work. This means you may need to rework this as a function to return the correct array and _not_ use `React.Fragment.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
-            onChange={listboxOnChange}
-            {...rest}
-          />
+          <>
```
You don't need this `<>` wrapper because the `Box` component is the only top level component.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,433 @@
+      };
+
+      const viewAllItems = () => {
+        setViewAllLabel(viewAllLabel === "View all" ? "View less" : "View all");
```
These `set....` functions are asynchronous and you're using the existing value to update the next value. To do this pass in a callback function:

```
setViewAllLabel((prevProp) => prevProp === "View all" ? "View less" : "View all");
```

Do this for all the other states where you use its own value to update it.

EdwinGuzman reviewed (changes requested):
It works well so far! I only reviewed the storybook file and the main `MultiSelect.tsx` component. I will review the rest next week but there are a few updates that should be made in the code before proceeding.

Commit:
fixed code review comments

Commit:
Fixed code review comments

Commit:
Fixed code review comments

Commit:
Fixed code review comments

Commit:
Fixed code review comments

Commit:
Fixed code review comments

Commit:
Fixed code review comments

Commit:
changed names in multiselect hook

Commit:
updated snapshot

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,528 @@
+| Component Version | DS Version |
+| ----------------- | ---------- |
+| Added             | `1.4.0`    |
+| Latest            | `2.0.0`    |
```
Can you update this to "Prerelease" instead of `2.0.0`

Commit:
changed names in multiselect hook

Commit:
changed names in multiselect hook

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,528 @@
+- {<Link href="#overview" target="_self">Overview</Link>}
+- {<Link href="#component-props" target="_self">Component Props</Link>}
+- {<Link href="#accessibility" target="_self">Accessibility</Link>}
+- {<Link href="#using-the-items-prop" target="_self">Using the items prop</Link>}
```
This is missing a line for "With Children Items Example".

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,528 @@
+/>
+
+## With Search Input Field Example
+When `isSearchable` is set to `true`, it enables the input field to search the items.
```
Put some extra line spaces between headings and paragraphs.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,528 @@
+  language="jsx"
+/>
+
+<Canvas of={MultiSelectStories.withDisabledAllChildrenItemsExample} />
```
Add a brief description that this example shows a parent and all its children as disabled.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,528 @@
+
+## With Default Open State Example
+When `isDefaultOpen` is set to `true`, the multi-select component is displayed in the open state by default.
+<Canvas of={MultiSelectStories.withSearchInputFieldExample} />
```
This example doesn't work. In the `withSearchInputFieldExample` component, make sure that `isDefaultOpen` is set to true.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -0,0 +1,437 @@
+      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=43593%3A24611",
+    },
+    jest: ["MultiSelect.test.tsx"],
+  },
```
Thanks for adding these for all stories but you don't need to. It's okay because the `withControls` component already has it.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -1,32 +1,33 @@
   {
     id: "colors",
     name: "Colors",
+    isdisabled: false,
```
This should be `isDisabled`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -1,32 +1,33 @@
 
+const defaultItemsVisible = 5;
+
 const MultiSelectTestDialogComponent = ({
```
Update this name because this is no longer the "dialog"

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -47,82 +48,48 @@ const MultiSelectTestDialogComponent = ({
-    expect(await axe(container)).toHaveNoViolations();
-  });
-
   it("should have no axe violations for the 'dialog' type", async () => {
```
Update the description. The "dialog" is no longer applicable.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -143,153 +110,134 @@ describe.skip("MultiSelect Dialog", () => {
-        labelText="MultiSelect Test Label"
-        type="dialog"
+        id="multiselect-test-id"
+        helperText="multiselect-helper-text"
+        buttonText="multiselect-button-text"
```
Gives these props better text and not id-style text.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -143,153 +110,134 @@ describe.skip("MultiSelect Dialog", () => {
-        name: /multiselect test label/i,
-      })
-    ).toBeInTheDocument();
+    expect(screen.getByRole("button")).toBeInTheDocument();
```
This is fine because it checks for the button, but the test description says it renders a given label and the label is not being tested here.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -143,153 +110,134 @@ describe.skip("MultiSelect Dialog", () => {
 
-  it("should render a 'Clear' and 'Apply button", () => {
+  it("should initially render with open menu if isDefaultOpen prop is true and when items are selected", () => {
+    selectedTestItems = { "multiselect-test-id": { items: ["red", "blue"] } };
```
For this test, you should verify that the checkboxes that has "red" and "blue" are checked.

bigfishdesign13 commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -0,0 +1,437 @@
+    },
+    onChange: {
+      description:
+        "The action to perform on the checkbox's onChange function. <br /> For dialog: `(event: React.ChangeEvent<HTMLInputElement>) => void` <br /> For listbox: `onChange: (selectedItem: MultiSelectItem, id: string) => void;`",
```
We no longer differentiate between `dialog` and `listbox`, so let's clean this up.  Only include the pattern that is valid.

bigfishdesign13 commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,528 @@
+`onMixedStateChange` for handling any changes to the selection of items and the
+current state of the selection: `selectedItems`.
+
+Find the full documentation under [useMultiSelect](../?path=/docs/hooks-usemultiselect--docs).
```
Is the `useMultiSelect` hook functional at this point?  If so, the story page should be enabled again.

bigfishdesign13 replied:
@gonuguntla  I see now that this was done, so ignore this comment.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -143,153 +110,134 @@ describe.skip("MultiSelect Dialog", () => {
+    ).not.toBeInTheDocument();
   });
 
   // Not sure this can be tested
```
Can you delete this commented out code block?

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -305,150 +253,117 @@ describe.skip("MultiSelect Dialog", () => {
-    expect(screen.queryByRole("checkbox")).toBeNull();
-    // TESTING FOR ENTER KEY
-    // Open multiselect using ENTER key.
+    // // TESTING FOR ENTER KEY
+    // // Open multiselect using ENTER key.
```
Minor: you don't need the double comment syntax.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -305,150 +253,117 @@ describe.skip("MultiSelect Dialog", () => {
-
-    // TESTING FOR SPACEBAR KEY
-    // Open multiselect using SPACEBAR key.
+    // // TESTING FOR SPACEBAR KEY
+    // // Open multiselect using SPACEBAR key.
```
Minor: you don't need the double comment syntax.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -305,150 +253,117 @@ describe.skip("MultiSelect Dialog", () => {
-    const onChangeMock = jest.fn();
-    const onMixedStateChangeMock = jest.fn();
 
     render(
```
In this case, try to use `rerender` because using `render` twice puts two `MultiSelect` components in the testing DOM while `rerender` will update the existing instance.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -459,24 +374,43 @@ describe.skip("MultiSelect Dialog", () => {
+      "true"
     );
 
+    render(
```
Also use `rerender` here.

bigfishdesign13 commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
+  /** The button text rendered within the MultiSelect. */
+  buttonText: string;
+  /** The selected items state (items that were checked by the user). */
+  defaultItemsVisible?: number;
```
Is this prop supposed to be working?  When I change in Storybook, I don't see a difference in the component.

gonuguntla replied:
Resolved this issue.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -520,149 +457,142 @@ describe.skip("MultiSelect Dialog", () => {
-    userEvent.click(screen.getByRole("button", { name: /MultiSelect Label/i }));
+    userEvent.click(screen.queryByRole("button"));
+
+    render(
```
Use rerender here, too.

EdwinGuzman replied:
Actually, you shouldn't need to render this component again here if you are clicking the button to open the component.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -520,149 +457,142 @@ describe.skip("MultiSelect Dialog", () => {
   });
 
   it("should render the UI snapshot correctly", () => {
     const defaultDialog = renderer
```
"Dialog" does not exist anymore.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
+  onMixedStateChange?: MultiSelectOnChange;
   /** The id of the MultiSelect. */
   id: string;
+  /** Boolean value used to enable the component's search functionality
```
This description is for `isSearchable`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
   /** Set the default open or closed state of the Multiselect. */
   isDefaultOpen?: boolean;
-  /** Boolean value used to control how the MultiSelect component will render within the page and interact with other DOM elements.
+  /** Boolean value used to control how the MultiSelect component will render input field within the page
```
This description is for `isBlockElement`.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
+
+      // Sets the selected items count on the menu button.
+      let getSelectedItemsCount: any;
+      getSelectedItemsCount = selectedItems[id]?.items.length || 0;
```
Since this never gets updated, make this:
`const getSelectedItemsCount: number = selectedItems[id]?.items.length || 0;`

EdwinGuzman replied:
Actually, `hasSelectedItems` is not being used in `multiSelect.ts` so you can remove this line.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
-        selectedItems,
-        type,
-        width,
+        hasSelectedItems: getSelectedItemsCount,
```
Because `getSelectedItemsCount` is a number but this is being used as a boolean for `hasSelectedItems`, update this to `!!getSelectedItemsCount` to get the boolean value of the variable.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
+              marginBottom="s"
+            />
+          )}
+          {!isSearchable && null}
```
What is this mean to do?

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
+
+      const showSearchInputBox = () => (
+        <>
+          {isSearchable && (
+            <TextInput
+              id="multi-select-text-input-id"
+              labelText={`Search ${buttonText}`}
+              isClearable={true}
+              isClearableCallback={clearSearchKeyword}
+              placeholder={`Search ${buttonText}`}
+              onChange={onChangeSearch}
+              showLabel={false}
+              showRequiredLabel={false}
+              type="text"
+              __css={styles.menuSearchInputBox}
+              marginBottom="s"
+            />
+          )}
```
If this is the only thing being returned from this function, then you don't need the react fragment `<>` wrapper.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,435 @@
+              name="multi-select-checkbox-group"
+            >
+              {itemsList.map((item: MultiSelectItem) => (
+                <>{generateCheckboxArray(item)}</>
```
This is really good, thanks.

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -26,8 +21,10 @@ const MultiSelect = {
     "menuButton",
     "menuContainer",
```
The following are not used so please remove: "actionButtons", "menu", "menuButton", "menuContainer"

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -38,6 +35,10 @@ const MultiSelect = {
       padding: { base: "s", md: "xs" },
       width: "auto",
     },
+    menuButton: {
```
This is not being used.

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -26,8 +21,10 @@ const MultiSelect = {
+  baseStyle: ({ isBlockElement, isOpen, width = "full" }) => ({
     ...multiSelectWidths[width],
     position: "relative",
     actionButtons: {
```
This is not being used so you can remove.

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -38,6 +35,10 @@ const MultiSelect = {
+      minWidth: width === "full" ? "100%" : "175px",
+      width: width === "full" ? "100%" : "fit-content",
+    },
     menuContainer: {
```
This is also not being used so remove

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -128,6 +129,60 @@ const MultiSelect = {
+      marginBottom: 0,
+      color: "ui.link.primary",
+    },
+    buttonTextLabel: {
```
Add this to the "parts" array.

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -128,6 +129,60 @@ const MultiSelect = {
+      }),
+      marginLeft: 54,
+    },
+    menuSearchInputBox: {
```
Add this to the "parts" array.

EdwinGuzman commented on src/theme/components/multiSelect.ts:
```diff
@@ -26,8 +21,10 @@ const MultiSelect = {
     "menuContainer",
     "menuChildren",
+    "viewAllButton",
+    "accordian",
```
This should be "accordionStyles" because that's the name of the object below.

EdwinGuzman reviewed (changes requested):
Some updates to clean some files. Can you also remove the `buttonLabel` and `menuButton` objects from the `multiSelectMenuButton.ts` file? Those are not being used in the new `MultiSelectMenuButton.tsx` file.

Commit:
updated some test cases

Commit:
Fixed code review comments

Commit:
Fixed code review comments

bigfishdesign13 committed:
cleanup of styles, docs, and functionality

Commit:
updated snapshot

Commit:
updated disabled checkbox test cases

Commit:
refactor code

bigfishdesign13 committed:
initial work for isBlockElement

Commit:
Updated useMultiSelectComponent hook documentation

Commit:
Updated useMultiSelectComponent hook documentation

bigfishdesign13 committed:
merge from DSD-1609/multiselect-v2

Commit:
updated changelog

bigfishdesign13 committed:
minor adjustments

bigfishdesign13 committed:
Merge branch 'DSD-1609/multiselect-v2-marty' into DSD-1609/multiselect-v2

Commit:
added code comments

Commit:
added code comments

bigfishdesign13 committed:
enabled isBlockElement

bigfishdesign13 committed:
Merge branch 'DSD-1609/multiselect-v2' of https://github.com/NYPL/nypl-design-system into DSD-1609/multiselect-v2

bigfishdesign13 commented on src/components/MultiSelect/MultiSelect.stories.tsx:
```diff
@@ -0,0 +1,548 @@
+      isDefaultOpen={isDefaultOpen}
+      isSearchable={isSearchable}
+      items={items}
+      listOverflow={listOverflow}
```
@EdwinGuzman I added this prop to allow for better control in the stories, but there is a type issue and I am not sure how to resolve.  PLease refer to the errors shown at the top of this file.

bigfishdesign13 committed:
updated theme file 'parts' list

bigfishdesign13 reviewed (changes requested):
I have some questions and I see one thing that needs to be addressed.

EdwinGuzman committed:
fixing type issue

bigfishdesign13 committed:
cleaning up stories

bigfishdesign13 committed:
Merge branch 'DSD-1609/multiselect-v2' of https://github.com/NYPL/nypl-design-system into DSD-1609/multiselect-v2

Commit:
Updated defaultdisplay values

bigfishdesign13 committed:
added a11y details

bigfishdesign13 committed:
Merge branch 'DSD-1609/multiselect-v2' of https://github.com/NYPL/nypl-design-system into DSD-1609/multiselect-v2

bigfishdesign13 committed:
added isBlockElement to isDefaultOpen example

bigfishdesign13 committed:
cleaning up dark mode styles

bigfishdesign13 committed:
added more story examples

Commit:
added defaultItemsVisible values in multi-select story

Commit:
added defaultItemsVisible values in multi-select story

Commit:
added defaultItemsVisible values in multi-select story

bigfishdesign13 committed:
added aria-label attribute to Menu main button

bigfishdesign13 commented:
@EdwinGuzman I was able to work out the `ariaLabel` prop on `Accordion`, which then allowed me to add a descriptive aria-label value to the `Menu` component's main button.  Can you please take a look at this latest update?  Thanks.

bigfishdesign13 committed:
formatting cleanup

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,716 @@
+
+## Accessibility
+
+The `MultiSelect` component is composed using the DS `Accordion` and `Checkbox`
```
I think linking to each storybook doc for Accordion and Checkbox is enough so we don't copy/paste the text. If the text in Accordion or Checkbox gets updated, we'd have to remember to update this  as well.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.mdx:
```diff
@@ -0,0 +1,716 @@
+When there are nested list items, only the parent items will be considered when
+counting the items to display.
+
+**IMPORTANT:** This prop can only be used with `listOverflow="expand"`.
```
I like the use of this as it stands out. Perhaps we can consider using the DS Notification to highlight these notes? Building the DS storybook with DS components -- an idea for later but just pointing out that this is readable already.

EdwinGuzman commented on src/components/Accordion/Accordion.test.tsx:
```diff
@@ -197,6 +210,69 @@ describe("Accordion", () => {
   });
 
+  describe("Check aria-label values", () => {
+    it("renders with attrbute generated from component prop", () => {
```
typo `attrbute`

EdwinGuzman replied:
Minor.

EdwinGuzman commented on src/components/Accordion/Accordion.test.tsx:
```diff
@@ -197,6 +210,69 @@ describe("Accordion", () => {
+      render(
+        <Accordion
+          accordionData={[accordionDataWithAriaLabel[0]]}
+          ariaLabel="Tom Nook, an Animal Crossing character"
```
Can you add an assertion that this global aria-label is _not_ in the document?

bigfishdesign13 replied:
I added a new test, but I needed to use a different method for testing that something does not exist.

EdwinGuzman commented on src/components/Accordion/Accordion.test.tsx:
```diff
@@ -197,6 +210,69 @@ describe("Accordion", () => {
+      );
+    });
+
+    it("generates warning ariaLabel is set twice", () => {
```
minor typo, add "when" before ariaLabel

EdwinGuzman commented on src/components/Accordion/Accordion.tsx:
```diff
@@ -14,12 +14,15 @@ import Icon from "../Icons/Icon";
 export interface AccordionProps {
   /** Array of data to display, and an optional accordionType */
   accordionData: AccordionDataProps[];
+  /** Description of aria-label. */
```
Can you add more description that this is a global default in case the data props don't have individual ariaLabel values?

EdwinGuzman commented on src/components/Accordion/Accordion.mdx:
```diff
@@ -2,15 +2,17 @@ import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";
+| Component Version | DS Version   |
+| ----------------- | ------------ |
+| Added             | `0.1.0`      |
+| Latest            | `Prerelease` |
```
This could use an update in the Accessibility section that adds information about the global ariaLabel prop and update the data related text that includes the data prop ariaLabel property in the object.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,474 @@
 export const MultiSelect = chakra(
   forwardRef<HTMLDivElement, React.PropsWithChildren<MultiSelectProps>>(
-    (props, ref?) => {
+    (props, _ref?) => {
```
Eventually this ref should go somewhere but I'm not sure where. This is not an issue now.

EdwinGuzman reviewed (commented):
Added comments about the recent set of updates. Otherwise, this looks very good!

One issue I can't pinpoint where it's happening, is when items are selected and the accordion button is focused, the selected item count button is hidden. When the focus is removed, the selected item button reappears.

https://github.com/NYPL/nypl-design-system/assets/1280564/5444213b-7d1f-4f60-9987-eb1d4e9c7521

Commit:
Added button ref for accordian component and when we close multi-select count focus on the main menu button

EdwinGuzman commented:
One other issue I noticed is the width of the list is sometimes narrower than the button which is not expected, right?

<img width="268" alt="Screen Shot 2024-01-30 at 9 13 03 AM" src="https://github.com/NYPL/nypl-design-system/assets/1280564/d6a72477-0a5a-4aed-98e6-77c290664caf">

EdwinGuzman commented on src/components/Accordion/Accordion.tsx:
```diff
@@ -32,6 +35,7 @@ export interface AccordionProps {
    * within accordion panel is greater than height set by panelMaxHeight, a
    * scrollbar will appear for accordion panel. */
   panelMaxHeight?: string;
+  buttonRef?: any;
```
This needs a comment to describe what this does.

EdwinGuzman commented on src/components/MultiSelect/MultiSelectMenuButton.tsx:
```diff
@@ -24,11 +19,8 @@ export interface MultiSelectMenuButtonProps {
   /** The action to perform for clear/reset button of MultiSelect. */
   onClear?: () => void;
   onKeyDown?: () => void;
+  buttonRef: any,
```
This component is a button as well so this could use a better prop name because it's confusing.

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,479 @@
 
-      const commonProps = {
-        id,
+       // Create a ref that we add to the element for which we want to detect outside clicks.
```
This description is not correct.

Commit:
fixed lint issues

EdwinGuzman commented on src/components/MultiSelect/MultiSelect.tsx:
```diff
@@ -1,138 +1,479 @@
-      const commonProps = {
-        id,
+       // Create a ref that we add to the element for which we want to detect outside clicks.
+       const internalRef: React.RefObject<HTMLDivElement> =
```
This can also use a better variable name because even though it is an internal ref, the way that it is used is very specific for the focus behavior.

EdwinGuzman reviewed (changes requested):
@gonuguntla the way `buttonRef` is implemented works, but the problem is that there is a single prop for this ref and it is being added to all accordion buttons. So if there are three accordion objects in the data array, they will all get the same ref and that's not ideal.
The pattern that was done for `ariaLabel` in each object can be used here. It's a bit tedious but better separates the ref for each accordion object. Try doing that instead. In a real use-case, the situation where a `MultiSelect` has more than one `Accordion` won't happen, but this should be built appropriately because it can use issues for other use-cases.

Commit:
Added better names for accordian buttonref

Commit:
Added better names for accordian buttonref

Added better names for accordian buttonref

Added better names for accordian buttonref

Added better names for accordian buttonref

add buttonref in accotdianData

add buttonref in accotdianData

added documentation

added documentation

Commit:
Added better names for accordian buttonref

Commit:
Added better names for accordian buttonref

Commit:
Added better names for accordian buttonref

bigfishdesign13 commented:
> One other issue I noticed is the width of the list is sometimes narrower than the button which is not expected, right?

@EdwinGuzman This is OK.  This is the expected behavior.

bigfishdesign13 commented:
> One issue I can't pinpoint where it's happening, is when items are selected and the accordion button is focused, the selected item count button is hidden. When the focus is removed, the selected item button reappears.

@EdwinGuzman  This is due to the z-index that is set in the focus styles.  The focus styles, by default, set the z-index to 9999, so the main button in the component is jumping over the counter button.  I was going to look into this today to see what could be done.

Commit:
add buttonref in accotdianData

Commit:
add buttonref in accotdianData

EdwinGuzman commented on CHANGELOG.md:
```diff
@@ -8,6 +8,15 @@ Currently, this repo is in Prerelease. When it is released, this project will ad
+
+### Updates
+
+- Updates the `Accordion` component to include the `ariaLabel` and `interactionRef` prop.
```
This should be `buttonInteractionRef`.

EdwinGuzman commented on src/components/Accordion/Accordion.mdx:
```diff
@@ -19,6 +21,7 @@ import Link from "../Link/Link";
 - {<Link href="#accessibility" target="_self">Accessibility</Link>}
 - {<Link href="#faq-example" target="_self">FAQ Example</Link>}
 - {<Link href="#example-with-panelmaxheight-prop" target="_self">Example with panelMaxHeight Prop</Link>}
+- {<Link href="#changelog" target="_self">Changelog</Link>}
```
In the "Overview" section, the second paragraph mentions the array of objects that contains `label`, `panel`, `accordionType`. This should also include the new `ariaLabel` and `buttonInteractionRef` properties.

bigfishdesign13 replied:
I added details about `ariaLabel`, but I will allow @gonuguntla to add details about `buttonInteractionRef`.

EdwinGuzman commented:
@gonuguntla the latest updates are good. It just needs to be explained a bit in the `Accordion` docs for the new property. Otherwise, Marty will look at the z-index issue.

I'm not sure this needs anything else right now but let's get another developer to help review the component.

bigfishdesign13 committed:
cleanup Accordion docs and tests

bigfishdesign13 committed:
z-index and organization of items count button

bigfishdesign13 committed:
merge from development branch

bigfishdesign13 commented on src/theme/components/multiSelectItemsCountButton.ts:
```diff
@@ -0,0 +1,39 @@
+    position: "absolute",
+    top: { base: "12px", md: "10px" },
+    width: "46px",
+    zIndex: 10000,
```
@EdwinGuzman I am not happy with this solution, but this is the only thing I could come up with to solve the layering issue.  Focus makes the main button jump to `9999`, so I simply set the "items count" button to `10000`.  Can you think of other possible solutions?

EdwinGuzman replied:
Oh, are we putting the 9999 value on focus? Let's leave this as 1000, it's not a problem but we should track all the z-index rules we have.

bigfishdesign13 replied:
Yes, `zIndex: "9999"` is part of the default focus styles.

bigfishdesign13 commented:
@EdwinGuzman and @gonuguntla – I renamed some of the files to reflect what they represented.  Also, I removed a lot of unnecessary styles that were originally for the button element in the v1 component.  We no longer need those styles becuase the button is now handled by the `Accordion` component.

bigfishdesign13 committed:
Accordion docs for buttonInteractionRef

bigfishdesign13 committed:
Merge branch 'development' into DSD-1609/multiselect-v2

jackiequach commented on src/components/Accordion/Accordion.test.tsx:
```diff
@@ -197,6 +210,77 @@ describe("Accordion", () => {
+      );
+    });
+
+    it("renders with attrbute generated from accordionData", () => {
```
Typo: attribute

jackiequach commented on src/components/Accordion/Accordion.tsx:
```diff
@@ -106,6 +113,16 @@ const getElementsFromData = (
+
+    if (content.ariaLabel && ariaLabel) {
+      console.warn(
+        "NYPL Reservoir Accordion: And arialLabel value has been passed for the " +
```
Typo, should be: An ariaLabel

jackiequach commented on src/hooks/useMultiSelect.mdx:
```diff
@@ -0,0 +1,107 @@
+| Latest       | `Prerelease`|
+
+`useMultiSelect` provides all functions needed to handle the selectedItems. This hook can be used for
+cases where managing the state of the component in the consuming app is less of a concern and general ease of use is prefered.
```
Typo: preferred

jackiequach reviewed (approved):
This looks great! Just some minor typos in docs.

bigfishdesign13 committed:
fixing typos and text wrapping

EdwinGuzman reviewed (approved):
Minor issue with the test but that's a typo issue, not a major issue.

Commit:
fixed spelling mistakes

bigfishdesign13 reviewed (approved):
Looks good and works well!

EdwinGuzman committed:
Merging from development
